File: Monopolya-desktop/GameModels.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace MonopolyWPF
{
    public enum CellType
    {
        Start,
        Property,
        Chance,
        CommunityChest,
        Jail,
        FreeParking,
        GoToJail,
        Tax,
        Lucky,
        Unlucky,
        Gift,
        Penalty
    }

    public class Cell
    {
        public string Name { get; set; }
        public CellType Type { get; set; }
        public int Position { get; set; }
        public Player? Owner { get; set; }

        public Cell(int position, string name, CellType type)
        {
            Position = position;
            Name = name;
            Type = type;
            Owner = null;
        }
    }

    public class Player
    {
        public string Name { get; set; }
        public int Position { get; set; }
        public int Money { get; set; }
        public bool InJail { get; set; }
        public int JailTurns { get; set; }
        public List<string> Properties { get; set; }

        public Player(string name)
        {
            Name = name;
            Position = 0;
            Money = 1500;
            InJail = false;
            JailTurns = 0;
            Properties = new List<string>();
        }

        public void Move(int steps, int boardSize)
        {
            if (boardSize <= 0) throw new ArgumentException("Board size must be positive");
            Position = (Position + steps) % boardSize;
        }
    }

    public class Dice
    {
        private Random random;

        public Dice()
        {
            random = new Random();
        }

        public int Roll()
        {
            return random.Next(1, 7);
        }

        public (int, int) RollTwoDice()
        {
            return (Roll(), Roll());
        }
    }

    public class MonopolyGame
    {
        private List<Player> players;
        private List<Cell> board;
        private Dice dice;
        private int currentPlayerIndex;
        private (int, int) lastDiceRoll;
        private string lastCardEffect = "";

        public List<Player> Players => players;
        public List<Cell> Board => board;

        public Player CurrentPlayer
        {
            get
            {
                if (players == null || players.Count == 0 || currentPlayerIndex >= players.Count)
                    throw new InvalidOperationException("No active players");
                return players[currentPlayerIndex];
            }
        }

        public Cell CurrentCell
        {
            get
            {
                var player = CurrentPlayer;
                if (board == null || player.Position >= board.Count)
                    throw new InvalidOperationException("Invalid player position");
                return board[player.Position];
            }
        }

        public (int, int) LastDiceRoll => lastDiceRoll;

        public MonopolyGame(int playerCount = 2, int startMoney = 1500)
        {
            dice = new Dice();
            players = new List<Player>();
            board = CreateBoard();
            currentPlayerIndex = 0;

            string[] colors = { "Красный", "Синий", "Зеленый", "Желтый" };
            for (int i = 0; i < Math.Min(playerCount, 4); i++)
            {
                var player = new Player(colors[i]);
                player.Money = startMoney;
                players.Add(player);
            }
        }

        private List<Cell> CreateBoard()
        {
            var board = new List<Cell>();

            // Верхний ряд (0-7) - разнообразные клетки
            board.Add(new Cell(0, "СТАРТ", CellType.Start));
            board.Add(new Cell(1, "Бульвар Арбат", CellType.Property));
            board.Add(new Cell(2, "Шанс", CellType.Chance));
            board.Add(new Cell(3, "Улица Тверская", CellType.Property));
            board.Add(new Cell(4, "Удача", CellType.Lucky));
            board.Add(new Cell(5, "Площадь Революции", CellType.Property));
            board.Add(new Cell(6, "Казна", CellType.CommunityChest));
            board.Add(new Cell(7, "Налог", CellType.Tax));

            // Правый ряд (8-14) - разнообразные клетки
            board.Add(new Cell(8, "ТЮРЬМА", CellType.Jail)); // Тюрьма в углу
            board.Add(new Cell(9, "Улица Пушкинская", CellType.Property));
            board.Add(new Cell(10, "Подарок", CellType.Gift));
            board.Add(new Cell(11, "Проспект Мира", CellType.Property));
            board.Add(new Cell(12, "Неудача", CellType.Unlucky));
            board.Add(new Cell(13, "Шанс", CellType.Chance));
            board.Add(new Cell(14, "Улица Чехова", CellType.Property));

            // Нижний ряд (15-22) - разнообразные клетки
            board.Add(new Cell(15, "Бесплатная стоянка", CellType.FreeParking));
            board.Add(new Cell(16, "Штраф", CellType.Penalty));
            board.Add(new Cell(17, "Улица Горького", CellType.Property));
            board.Add(new Cell(18, "Казна", CellType.CommunityChest));
            board.Add(new Cell(19, "Удача", CellType.Lucky));
            board.Add(new Cell(20, "Вокзал", CellType.Property));
            board.Add(new Cell(21, "Налог на роскошь", CellType.Tax));
            board.Add(new Cell(22, "Подарок", CellType.Gift));

            // Левый ряд (23-28) - разнообразные клетки
            board.Add(new Cell(23, "Отправляйтесь в тюрьму", CellType.GoToJail));
            board.Add(new Cell(24, "Улица Ленина", CellType.Property));
            board.Add(new Cell(25, "Неудача", CellType.Unlucky));
            board.Add(new Cell(26, "Проспект Победы", CellType.Property));
            board.Add(new Cell(27, "Штраф", CellType.Penalty));
            board.Add(new Cell(28, "Шанс", CellType.Chance));

            return board;
        }

        public void StartNewGame()
        {
            foreach (var player in players)
            {
                player.Position = 0;
                player.InJail = false;
                player.JailTurns = 0;
                player.Properties.Clear();
            }

            foreach (var cell in board)
            {
                cell.Owner = null;
            }

            currentPlayerIndex = 0;
        }

        public void RollDice()
        {
            lastDiceRoll = dice.RollTwoDice();
            lastCardEffect = "";
        }

        public void MoveCurrentPlayer()
        {
            if (players.Count == 0) return;

            int total = lastDiceRoll.Item1 + lastDiceRoll.Item2;

            if (CurrentPlayer.InJail)
            {
                HandleJailTurn();
                return;
            }

            int oldPosition = CurrentPlayer.Position;
            CurrentPlayer.Move(total, board.Count);

            if (oldPosition + total >= board.Count)
            {
                CurrentPlayer.Money += 200;
            }
        }

        public void HandleCurrentCellAction()
        {
            HandleCellAction(CurrentCell);
        }

        private void HandleJailTurn()
        {
            CurrentPlayer.JailTurns--;

            if (lastDiceRoll.Item1 == lastDiceRoll.Item2)
            {
                CurrentPlayer.InJail = false;
                CurrentPlayer.JailTurns = 0;

                int total = lastDiceRoll.Item1 + lastDiceRoll.Item2;
                int oldPosition = CurrentPlayer.Position;
                CurrentPlayer.Move(total, board.Count);

                if (oldPosition + total >= board.Count)
                {
                    CurrentPlayer.Money += 200;
                }
            }
            else if (CurrentPlayer.JailTurns <= 0)
            {
                CurrentPlayer.Money -= 50;
                CurrentPlayer.InJail = false;

                int total = lastDiceRoll.Item1 + lastDiceRoll.Item2;
                int oldPosition = CurrentPlayer.Position;
                CurrentPlayer.Move(total, board.Count);

                if (oldPosition + total >= board.Count)
                {
                    CurrentPlayer.Money += 200;
                }
            }
        }

        private void HandleCellAction(Cell cell)
        {
            if (cell == null) return;

            switch (cell.Type)
            {
                case CellType.Chance:
                    DrawChanceCard();
                    break;

                case CellType.CommunityChest:
                    DrawCommunityChestCard();
                    break;

                case CellType.Lucky:
                    int luckAmount = new Random().Next(50, 201);
                    CurrentPlayer.Money += luckAmount;
                    lastCardEffect = $"Удача! +${luckAmount}";
                    break;

                case CellType.Unlucky:
                    int unluckAmount = new Random().Next(30, 151);
                    CurrentPlayer.Money -= unluckAmount;
                    lastCardEffect = $"Неудача! -${unluckAmount}";
                    break;

                case CellType.Gift:
                    int giftAmount = new Random().Next(40, 121);
                    CurrentPlayer.Money += giftAmount;
                    lastCardEffect = $"Подарок! +${giftAmount}";
                    break;

                case CellType.Penalty:
                    int penaltyAmount = new Random().Next(20, 101);
                    CurrentPlayer.Money -= penaltyAmount;
                    lastCardEffect = $"Штраф! -${penaltyAmount}";
                    break;

                case CellType.GoToJail:
                    CurrentPlayer.Position = 8; // Позиция тюрьмы
                    CurrentPlayer.InJail = true;
                    CurrentPlayer.JailTurns = 3;
                    lastCardEffect = "Отправляйтесь в тюрьму!";
                    break;

                case CellType.Tax:
                    int tax = cell.Name.Contains("роскошь") ? 100 : 50;
                    CurrentPlayer.Money -= tax;
                    lastCardEffect = $"Налог -${tax}";
                    break;
            }
        }

        private void DrawChanceCard()
        {
            string[] chances = {
                "Аванс на работе! Получите $100",
                "Штраф за превышение скорости. Заплатите $50",
                "Вы выиграли в лотерею! Получите $150",
                "Ремонт дома. Заплатите $80",
                "Перейдите на СТАРТ",
                "Отправляйтесь в тюрьму",
                "Получите $50 от каждого игрока"
            };

            Random rand = new Random();
            string chance = chances[rand.Next(chances.Length)];
            ApplyChanceEffect(chance);
            lastCardEffect = chance;
        }

        private void ApplyChanceEffect(string chance)
        {
            if (string.IsNullOrEmpty(chance)) return;

            if (chance.Contains("Получите $100"))
                CurrentPlayer.Money += 100;
            else if (chance.Contains("Заплатите $50"))
                CurrentPlayer.Money -= 50;
            else if (chance.Contains("Получите $150"))
                CurrentPlayer.Money += 150;
            else if (chance.Contains("Заплатите $80"))
                CurrentPlayer.Money -= 80;
            else if (chance.Contains("Перейдите на СТАРТ"))
            {
                CurrentPlayer.Position = 0;
                CurrentPlayer.Money += 200;
            }
            else if (chance.Contains("Отправляйтесь в тюрьму"))
            {
                CurrentPlayer.Position = 8;
                CurrentPlayer.InJail = true;
                CurrentPlayer.JailTurns = 3;
            }
            else if (chance.Contains("Получите $50 от каждого игрока"))
            {
                foreach (var p in players.Where(p => p != CurrentPlayer && p.Money >= 50))
                {
                    p.Money -= 50;
                    CurrentPlayer.Money += 50;
                }
            }
        }

        private void DrawCommunityChestCard()
        {
            string[] chests = {
                "Наследство! Получите $100",
                "Оплатите лечение. Заплатите $50",
                "День рождения! Получите $10 от каждого игрока",
                "Налоговая проверка. Заплатите $75",
                "Возврат налогов. Получите $20",
                "Благотворительность. Заплатите $30"
            };

            Random rand = new Random();
            string chest = chests[rand.Next(chests.Length)];
            ApplyChestEffect(chest);
            lastCardEffect = chest;
        }

        private void ApplyChestEffect(string chest)
        {
            if (string.IsNullOrEmpty(chest)) return;

            if (chest.Contains("Получите $100"))
                CurrentPlayer.Money += 100;
            else if (chest.Contains("Заплатите $50"))
                CurrentPlayer.Money -= 50;
            else if (chest.Contains("Получите $10 от каждого игрока"))
            {
                foreach (var p in players.Where(p => p != CurrentPlayer && p.Money >= 10))
                {
                    p.Money -= 10;
                    CurrentPlayer.Money += 10;
                }
            }
            else if (chest.Contains("Заплатите $75"))
                CurrentPlayer.Money -= 75;
            else if (chest.Contains("Получите $20"))
                CurrentPlayer.Money += 20;
            else if (chest.Contains("Заплатите $30"))
                CurrentPlayer.Money -= 30;
        }

        public string GetLastCardEffect()
        {
            return lastCardEffect;
        }

        public bool BuyCurrentProperty()
        {
            var cell = CurrentCell;
            if (cell.Type == CellType.Property && cell.Owner == null)
            {
                int price = 100 + cell.Position * 15;
                if (CurrentPlayer.Money >= price)
                {
                    CurrentPlayer.Money -= price;
                    CurrentPlayer.Properties.Add(cell.Name);
                    cell.Owner = CurrentPlayer;
                    return true;
                }
            }
            return false;
        }

        public void NextPlayer()
        {
            if (players.Count > 0)
            {
                currentPlayerIndex = (currentPlayerIndex + 1) % players.Count;
            }
        }

        public bool CheckGameOver()
        {
            return players.Any(p => p.Money <= 0);
        }
    }
}

File: Monopolya-mobile/app/src/main/java/com/example/monopolya/GameModels.kt
package com.example.monopolya

import androidx.compose.ui.graphics.Color
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlin.random.Random

enum class CellType { START, PROPERTY, CHANCE, COMMUNITY_CHEST, JAIL, FREE_PARKING, GO_TO_JAIL, TAX, LUCKY, UNLUCKY, GIFT, PENALTY }

data class Cell(val position: Int, val name: String, val type: CellType, var owner: Player? = null)
data class Player(val name: String, var position: Int = 0, var money: Int = 1500, var inJail: Boolean = false, var jailTurns: Int = 0, val properties: MutableList<String> = mutableListOf())

class MonopolyGame(playerCount: Int = 2, startMoney: Int = 1500) {
    private val _players = MutableStateFlow<List<Player>>(emptyList())
    val players: StateFlow<List<Player>> = _players.asStateFlow()
    private val _board = MutableStateFlow<List<Cell>>(emptyList())
    val board: StateFlow<List<Cell>> = _board.asStateFlow()
    private val _currentPlayerIndex = MutableStateFlow(0)
    val currentPlayerIndex: StateFlow<Int> = _currentPlayerIndex.asStateFlow()
    private val _lastDiceRoll = MutableStateFlow(Pair(1, 1))
    val lastDiceRoll: StateFlow<Pair<Int, Int>> = _lastDiceRoll.asStateFlow()
    private val _lastCardEffect = MutableStateFlow("")
    val lastCardEffect: StateFlow<String> = _lastCardEffect.asStateFlow()
    private val _isMoving = MutableStateFlow(false)
    val isMoving: StateFlow<Boolean> = _isMoving.asStateFlow()
    private val dice = Random(System.currentTimeMillis())
    val currentPlayer: Player get() = _players.value[_currentPlayerIndex.value]
    val currentCell: Cell get() = _board.value[currentPlayer.position]

    init {
        _board.value = listOf(
            Cell(0, "СТАРТ", CellType.START), Cell(1, "Бульвар Арбат", CellType.PROPERTY),
            Cell(2, "Шанс", CellType.CHANCE), Cell(3, "Улица Тверская", CellType.PROPERTY),
            Cell(4, "Удача", CellType.LUCKY), Cell(5, "Площадь Революции", CellType.PROPERTY),
            Cell(6, "Казна", CellType.COMMUNITY_CHEST), Cell(7, "Налог", CellType.TAX),
            Cell(8, "ТЮРЬМА", CellType.JAIL), Cell(9, "Улица Пушкинская", CellType.PROPERTY),
            Cell(10, "Подарок", CellType.GIFT), Cell(11, "Проспект Мира", CellType.PROPERTY),
            Cell(12, "Неудача", CellType.UNLUCKY), Cell(13, "Шанс", CellType.CHANCE),
            Cell(14, "Улица Чехова", CellType.PROPERTY), Cell(15, "Бесплатная стоянка", CellType.FREE_PARKING),
            Cell(16, "Штраф", CellType.PENALTY), Cell(17, "Улица Горького", CellType.PROPERTY),
            Cell(18, "Казна", CellType.COMMUNITY_CHEST), Cell(19, "Удача", CellType.LUCKY),
            Cell(20, "Вокзал", CellType.PROPERTY), Cell(21, "Налог на роскошь", CellType.TAX),
            Cell(22, "Подарок", CellType.GIFT), Cell(23, "Отправляйтесь в тюрьму", CellType.GO_TO_JAIL),
            Cell(24, "Улица Ленина", CellType.PROPERTY), Cell(25, "Неудача", CellType.UNLUCKY),
            Cell(26, "Проспект Победы", CellType.PROPERTY), Cell(27, "Штраф", CellType.PENALTY),
            Cell(28, "Шанс", CellType.CHANCE)
        )
        val playersList = mutableListOf<Player>()
        val colors = listOf("Красный", "Синий", "Зеленый", "Желтый")
        for (i in 0 until minOf(playerCount, 4)) playersList.add(Player(colors[i], money = startMoney))
        _players.value = playersList
    }

    fun startNewGame() {
        _players.value = _players.value.map { it.apply { position = 0; inJail = false; jailTurns = 0; properties.clear() } }
        _board.value = _board.value.map { it.apply { owner = null } }
        _currentPlayerIndex.value = 0
    }

    fun rollDice() { _lastDiceRoll.value = Pair(dice.nextInt(1, 7), dice.nextInt(1, 7)) }

    suspend fun moveCurrentPlayer(onStepComplete: (Int) -> Unit) {
        val total = _lastDiceRoll.value.first + _lastDiceRoll.value.second
        if (currentPlayer.inJail) { handleJailTurn(total); return }
        val oldPosition = currentPlayer.position
        _isMoving.value = true
        for (step in 1..total) {
            val stepPosition = (oldPosition + step) % _board.value.size
            _players.value.toMutableList().apply { this[_currentPlayerIndex.value].position = stepPosition }.let { _players.value = it }
            onStepComplete(stepPosition)
            kotlinx.coroutines.delay(150)
        }
        _isMoving.value = false
        if (oldPosition + total >= _board.value.size) {
            _players.value.toMutableList().apply { this[_currentPlayerIndex.value].money += 200 }.let { _players.value = it }
            _lastCardEffect.value = "Проход через СТАРТ! +$200"
        }
        handleCurrentCellAction()
    }

    private fun handleJailTurn(total: Int) {
        val playersList = _players.value.toMutableList()
        val player = playersList[_currentPlayerIndex.value]
        player.jailTurns--
        if (_lastDiceRoll.value.first == _lastDiceRoll.value.second) {
            player.inJail = false; player.jailTurns = 0
            player.position = (player.position + total) % _board.value.size
            _lastCardEffect.value = "Вы вышли из тюрьмы!"
        } else if (player.jailTurns <= 0) {
            player.money -= 50; player.inJail = false
            player.position = (player.position + total) % _board.value.size
            _lastCardEffect.value = "Оплачено $50 за выход из тюрьмы"
        } else _lastCardEffect.value = "В тюрьме! Осталось ходов: ${player.jailTurns}"
        _players.value = playersList
    }

    private fun handleCurrentCellAction() {
        val cell = currentCell
        val playersList = _players.value.toMutableList()
        val player = playersList[_currentPlayerIndex.value]
        when (cell.type) {
            CellType.CHANCE -> drawCard(true)
            CellType.COMMUNITY_CHEST -> drawCard(false)
            CellType.LUCKY -> { val a = Random.nextInt(50, 201); player.money += a; _lastCardEffect.value = "Удача! +$$a" }
            CellType.UNLUCKY -> { val a = Random.nextInt(30, 151); player.money -= a; _lastCardEffect.value = "Неудача! -$$a" }
            CellType.GIFT -> { val a = Random.nextInt(40, 121); player.money += a; _lastCardEffect.value = "Подарок! +$$a" }
            CellType.PENALTY -> { val a = Random.nextInt(20, 101); player.money -= a; _lastCardEffect.value = "Штраф! -$$a" }
            CellType.GO_TO_JAIL -> { player.position = 8; player.inJail = true; player.jailTurns = 3; _lastCardEffect.value = "Отправляйтесь в тюрьму!" }
            CellType.TAX -> { val tax = if (cell.name.contains("роскошь")) 100 else 50; player.money -= tax; _lastCardEffect.value = "Налог -$$tax" }
            else -> {}
        }
        _players.value = playersList
    }

    private fun drawCard(isChance: Boolean) {
        val effects = if (isChance) listOf(
            "Аванс! +$100" to { currentPlayer.money += 100 },
            "Штраф! -$50" to { currentPlayer.money -= 50 },
            "Лотерея! +$150" to { currentPlayer.money += 150 },
            "Ремонт! -$80" to { currentPlayer.money -= 80 },
            "На СТАРТ! +$200" to { currentPlayer.position = 0; currentPlayer.money += 200 },
            "В тюрьму!" to { currentPlayer.position = 8; currentPlayer.inJail = true; currentPlayer.jailTurns = 3 }
        ) else listOf(
            "Наследство! +$100" to { currentPlayer.money += 100 },
            "Лечение! -$50" to { currentPlayer.money -= 50 },
            "День рождения! +$10 от всех" to { _players.value.filter { it != currentPlayer && it.money >= 10 }.forEach { it.money -= 10; currentPlayer.money += 10 } },
            "Налоги! -$75" to { currentPlayer.money -= 75 },
            "Возврат! +$20" to { currentPlayer.money += 20 },
            "Благотворительность! -$30" to { currentPlayer.money -= 30 }
        )
        val (effect, action) = effects.random()
        action()
        _lastCardEffect.value = effect
    }

    fun buyCurrentProperty(): Boolean {
        val cell = currentCell
        if (cell.type == CellType.PROPERTY && cell.owner == null) {
            val price = 100 + cell.position * 15
            if (currentPlayer.money >= price) {
                _players.value.toMutableList().apply { this[_currentPlayerIndex.value].money -= price; this[_currentPlayerIndex.value].properties.add(cell.name) }.let { _players.value = it }
                _board.value.toMutableList().apply { this[cell.position].owner = currentPlayer }.let { _board.value = it }
                _lastCardEffect.value = "Куплено: ${cell.name} за $$price"
                return true
            }
        }
        return false
    }

    fun nextPlayer() { if (_players.value.isNotEmpty()) _currentPlayerIndex.value = (_currentPlayerIndex.value + 1) % _players.value.size }
    fun checkGameOver(): Boolean = _players.value.any { it.money <= 0 }
    fun getCellPrice(position: Int): Int = 100 + position * 15
    fun getCellIcon(type: CellType) = when (type) {
        CellType.START -> "🏁"; CellType.PROPERTY -> "🏘️"; CellType.CHANCE -> "❓"
        CellType.COMMUNITY_CHEST -> "📦"; CellType.JAIL -> "⛓️"; CellType.FREE_PARKING -> "🅿️"
        CellType.GO_TO_JAIL -> "🚓"; CellType.TAX -> "💰"; CellType.LUCKY -> "🍀"
        CellType.UNLUCKY -> "💢"; CellType.GIFT -> "🎁"; CellType.PENALTY -> "⚠️"
        else -> "⬜"
    }
    fun getPlayerColor(index: Int) = when (index % 4) { 0 -> Color(0xFFFF5050); 1 -> Color(0xFF5050FF); 2 -> Color(0xFF50FF50); 3 -> Color(0xFFFFFF50); else -> Color.Gray }
    fun getCellColor(position: Int): Color {
        val cell = _board.value.getOrNull(position) ?: return Color.Gray
        if (cell.owner != null) return getPlayerColor(_players.value.indexOf(cell.owner))
        return when (cell.type) {
            CellType.START -> Color(0xFFC8FFC8); CellType.PROPERTY -> Color(0xFFC8DCFF); CellType.CHANCE -> Color(0xFFFFC896)
            CellType.COMMUNITY_CHEST -> Color(0xFFFFFFC8); CellType.JAIL -> Color(0xFFFF9696); CellType.FREE_PARKING -> Color(0xFFF0F0F0)
            CellType.GO_TO_JAIL -> Color(0xFFFF6464); CellType.TAX -> Color(0xFFFFC8DC); CellType.LUCKY -> Color(0xFFFFD700)
            CellType.UNLUCKY -> Color(0xFFA9A9A9); CellType.GIFT -> Color(0xFFEE82EE); CellType.PENALTY -> Color(0xFFFF8C00)
            else -> Color.White
        }
    }
    fun getDiceSymbol(value: Int) = when (value) { 1 -> "⚀"; 2 -> "⚁"; 3 -> "⚂"; 4 -> "⚃"; 5 -> "⚄"; 6 -> "⚅"; else -> "⚀" }
}